### DISCUSSION TITLE
Impossible to target a chosen page on a multipage Inkscape document

### DISCUSSION THREAD

**Discussion by duangdoock** [spakin/SimpInkScr#140] [Q&A] [answered]
In Inkscape when multipages are created it's impossible to target the script to the desired page. It is always rendered on the first page only. So how would it be possible to target the script on page 2 only for example? Thanks

**Comment by spakin** [comment]
Because SVG doesn't actually support pages, Inkscape implements them as coordinates that lie outside of the viewbox.  Simple Inkscape Scripting enables scripts to query a page's coordinates by invoking the `bounding_box` method on the page object, as in the following example:
```Python
# Create a few pages.
npages = 3
pages = [page(i + 1) for i in range(npages)]

# Draw a large circle in the center of each page.
rad = min(pages[0].width, pages[0].height)*3/8
for pg in pages:
    bbox = pg.bounding_box()
    circle(bbox.center, rad, stroke=2, fill=randcolor())
```

Hence, to target a script to a particular page, you'll need to acquire the page's bounding box as in the preceding example and limit the script's operations only to objects that lie within that bounding box.

See [Page creation](https://github.com/spakin/SimpInkScr/wiki/Document-layout#page-creation) in the Simple Inkscape Scripting wiki for more information.

**Reply by duangdoock** [reply]
Thanks very much for taking the time to reply.
In fact I would like to automatically insert images in an Inkscape file which already has multipages.

Exemple, on page 1:
Image1.png at position x=200px y=200px
Image2.png at position x=400px y=400px

On page 2:
Image3.png at position x=200px y=200px
Image4.png at position x=400px y=400px

Do you think that a script could do it? 
Thanks in advance for your help.

**Reply by spakin** [reply, marked as answer]
Yes.  You can acquire the bounding box for each page then add each bounding box's left coordinate to _x_ and top coordinate to _y_:
```Python
bboxes = [pg.bounding_box() for pg in all_pages()]
image('Image1.png', (bboxes[0].left + 200, bboxes[0].top + 200))
image('Image2.png', (bboxes[0].left + 400, bboxes[0].top + 400))
image('Image3.png', (bboxes[1].left + 200, bboxes[1].top + 200))
image('Image4.png', (bboxes[1].left + 400, bboxes[1].top + 400))
```
Alternatively, you might want to use an [`inkex.transforms.ImmutableVector2d`](https://inkscape.gitlab.io/extensions/documentation/_modules/inkex/transforms.html#ImmutableVector2d) to make the code easier to read:
```Python
bboxes = [pg.bounding_box() for pg in all_pages()]
origins = [inkex.transforms.ImmutableVector2d(bb.left, bb.top) for bb in bboxes]
image('Image1.png', origins[0] + (200, 200))
image('Image2.png', origins[0] + (400, 400))
image('Image3.png', origins[1] + (200, 200))
image('Image4.png', origins[1] + (400, 400))
```

**Reply by duangdoock** [reply]
Excellent Spakin you solved exactly my problem. It's working great! Thanks for the effort you're putting in this useful extension. 
All the best!

**Reply by duangdoock** [reply]
If I may, how can I add a "rotation argument" to each pictures following the 2 options you suggested?

**Reply by spakin** [reply]
Invoke the `rotate` method on the object returned by `image`: `image(…).rotate(…)`.  For example
```Python
image('Image1.png', origins[0] + (200, 200)).rotate(45)
```
See [Applying individual transformations](https://github.com/spakin/SimpInkScr/wiki/Modifying-existing-objects#applying-individual-transformations) in the [Simple Inkscape Scripting wiki](https://github.com/spakin/SimpInkScr/wiki) for more information about `rotate` and similar transformations.

**Reply by duangdoock** [reply]
Thanks so much!

**Reply by duangdoock** [reply]
Hello
I'm requesting your help for the rotation again.
I've noticed that the "image(…).rotate(…)" argument only rotate the image around its center. I would be interested to rotate the image around its upper-left corner for example, or around an (x, y) coordinates. Is it possible? 
Thanks in advance. 
Kind regards

**Reply by spakin** [reply]
Yes.  Again, read [Applying individual transformations](https://github.com/spakin/SimpInkScr/wiki/Modifying-existing-objects#applying-individual-transformations).  That page describes how `rotate` accepts an `anchor` argument that rotates around either a named position on the shape or an absolute (_x_, _y_) coordinate on the page.  In particular, to rotate around the image's upper-left corner, use `image(…).rotate(…, anchor="ul")`.

**Reply by duangdoock** [reply]
Thanks very much Spakin for these tips. In my case **image(…).rotate(…, anchor="ul")** returned me an error. But as suggested through the link you sent me I replaced it with **image(…).rotate(…, 'ul')** with success. 🙏
All the best!

**Reply by spakin** [reply]
It looks like there was a mismatch between the documentation and the code.  The former called the argument "anchor" when it was in fact named "around" in the code.  In a605d85183dd60568fe744b87746903b7722ee32 I modified the code to match the documentation so `image(…).rotate(…, anchor="ul")` should also work now.

**Reply by duangdoock** [reply]
Thanks for this information. All the best!
